File: java/ru/eustas/twim/SinCos.java
package ru.eustas.twim;

class SinCos {
  static final int SCALE = 1 << 18;

  static final int MAX_ANGLE_BITS = 9;
  static final int MAX_ANGLE = 1 << MAX_ANGLE_BITS;

  static final int[] SIN = new int[MAX_ANGLE];
  static final int[] COS = new int[MAX_ANGLE];
  static final double[] INV_SIN = new double[MAX_ANGLE];
  static final float[] MINUS_COT = new float[MAX_ANGLE];

  static {
    for (int i = 0; i < MAX_ANGLE; ++i) {
      SIN[i] = (int) Math.round(SCALE * Math.sin((Math.PI * i) / MAX_ANGLE));
      COS[i] = (int) Math.round(SCALE * Math.cos((Math.PI * i) / MAX_ANGLE));
      INV_SIN[i] = (i != 0) ? (1.0 / SIN[i]) : 0.0;
      MINUS_COT[i] = (float) (-COS[i] * INV_SIN[i]);
    }
  }
}

File: js/SinCos.js
import {b8, b16, math, mathRound, newInt32Array} from "./Mini.js";

export let /** @noinline @type{number} */ SCALE = b16 * 4;
export const /** @type{number} */ MAX_ANGLE_BITS = 9;
export let /** @noinline @type{number} */ MAX_ANGLE = b8 * 2;

export let /** @const @type{!Int32Array} */ SIN = newInt32Array(MAX_ANGLE);
export let /** @const @type{!Int32Array} */ COS = newInt32Array(MAX_ANGLE);

for (let /** @type{number} */ i = 0; i < MAX_ANGLE; ++i) {
  let /** @type{number} */ angle = (math.PI * i) / MAX_ANGLE;
  SIN[i] = mathRound(SCALE * math.sin(angle));
  COS[i] = mathRound(SCALE * math.cos(angle));
}
